homeassistant/components/control4/media_player.py

```diff
@@ -0,0 +1,406 @@
+
+        if source_state := self._get_current_source_state():
+            return source_state
+
+        if self.coordinator.data[self._idx][CONTROL4_POWER_STATE]:
+            return MediaPlayerState.ON
+
+        if self._is_soft_on:
+            return MediaPlayerState.IDLE
+
+        return MediaPlayerState.OFF
+
+    @property
+    def source(self):
+        """Get the current source."""
+        current_source = self._get_current_playing_device_id()
+        if not current_source or current_source not in self._sources:
+            return None
+        return self._sources[current_source].name
+
+    @property
+    def media_title(self) -> str | None:
+        """Get the Media Title."""
+        media_info = self._get_media_info()
+        if not media_info:
+            return None
+        if "title" in media_info:
+            return media_info["title"]
+        current_source = self._get_current_playing_device_id()
+        if not current_source or current_source not in self._sources:
+            return None
+        return self._sources[current_source].name
+
+    @property
+    def media_content_type(self):
+        """Get current content type if available."""
+        current_source = self._get_current_playing_device_id()
+        if not current_source:
+            return None
+        if current_source == self._get_current_video_device_id():
+            return MediaType.VIDEO
+        return MediaType.MUSIC
+
+    async def async_media_play_pause(self):
+        """If possible, toggle the current play/pause state.
+
+        Not every source supports play/pause.
+        Unfortunately MediaPlayer capabilities are not dynamic,
+        so we must determine if play/pause is supported here
+        """
+        if self._get_current_source_state():
+            await super().async_media_play_pause()
+
+    @property
+    def source_list(self) -> list[str]:
+        """Get the available source."""
+        return [x.name for x in self._sources.values()]
+
+    @property
+    def volume_level(self):
+        """Get the volume level."""
+        return self.coordinator.data[self._idx][CONTROL4_VOLUME_STATE] / 100
+
+    @property
+    def is_volume_muted(self):
+        """Check if the volume is muted."""
+        return bool(self.coordinator.data[self._idx][CONTROL4_MUTED_STATE])
+
+    async def async_select_source(self, source):
+        """Select a new source."""
+        for avail_source in self._sources.values():
+            if avail_source.name == source:
+                audio_only = _SourceType.VIDEO not in avail_source.source_type
+                if audio_only:
+                    await self._create_api_object().setAudioSource(avail_source.id)
+                else:
+                    await self._create_api_object().setVideoAndAudioSource(
+                        avail_source.id
+                    )
+                break
+
+        await self.coordinator.async_request_refresh()
+
+    def turn_on(self):
```

Comment:
This is weird. How about setting the feature flag for `TURN_OFF` but not for `TURN_ON`, and after a call to `TURN_OFF` set the state to `MediaPlayerState.IDLE` or `MediaPlayerState.STANDBY` instead of `MediaPlayerState.OFF`?

Reply:
Yeah we can do that if we don't want to reflect the true state of off or on. Since in control4 by playing a source turns on the room and sets media

Reply:
Oh yeah. The UI breaks in many cases as we cannot start playing without the device being ON. Hence the necessity for this fake turn on.

Reply:
Done off sets state to idle no more soft on/fake off.